library/objective-c/EnvoyHTTPStreamImpl.m

```diff
@@ -126,50 +126,59 @@ static void ios_on_complete(void *context) {
   ios_context *c = (ios_context *)context;
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
   dispatch_async(callbacks.dispatchQueue, ^{
-    // TODO: release stream
-    if (atomic_load(c->canceled)) {
+    if (atomic_load(c->canceled) || !callbacks._onStreamClose) {
       return;
     }
+    callbacks._onStreamClose();
   });
 }
 
 static void ios_on_cancel(void *context) {
   ios_context *c = (ios_context *)context;
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
-  // TODO: release stream
   dispatch_async(callbacks.dispatchQueue, ^{
     // This call is atomically gated at the call-site and will only happen once.
-    if (!callbacks.onCancel) {
-      return;
+    if (callbacks.onCancel) {
+      callbacks.onCancel();
+    }
+    if (callbacks._onStreamClose) {
+      callbacks._onStreamClose();
     }
-    callbacks.onCancel();
   });
 }
 
 static void ios_on_error(envoy_error error, void *context) {
   ios_context *c = (ios_context *)context;
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
   dispatch_async(callbacks.dispatchQueue, ^{
-    // TODO: release stream
-    if (atomic_load(c->canceled) || !callbacks.onError) {
+    if (atomic_load(c->canceled)) {
       return;
     }
-    NSString *errorMessage = [[NSString alloc] initWithBytes:error.message.bytes
-                                                      length:error.message.length
-                                                    encoding:NSUTF8StringEncoding];
-    error.message.release(error.message.context);
-    callbacks.onError(error.error_code, errorMessage);
+    if (callbacks.onError) {
+      NSString *errorMessage = [[NSString alloc] initWithBytes:error.message.bytes
+                                                        length:error.message.length
+                                                      encoding:NSUTF8StringEncoding];
+      error.message.release(error.message.context);
+      callbacks.onError(error.error_code, errorMessage);
+    }
+    if (callbacks._onStreamClose) {
+      callbacks._onStreamClose();
+    }
   });
 }
 
 #pragma mark - EnvoyHTTPStreamImpl
 
-@implementation EnvoyHTTPStreamImpl {
-  EnvoyHTTPStreamImpl *_strongSelf;
-  EnvoyHTTPCallbacks *_platformCallbacks;
-  envoy_http_callbacks _nativeCallbacks;
-  envoy_stream_t _streamHandle;
-}
+@interface EnvoyHTTPStreamImpl ()
+
+@property (nonatomic, strong) EnvoyHTTPStreamImpl *_strongSelf;
+@property (nonatomic, strong) EnvoyHTTPCallbacks *_platformCallbacks;
```

Comment:
Will dealloc get rid of this strong ref automatically? Or do we need to assign nil explicitly?

Reply:
ivars use `__strong` references for objects by default even though it's not specified explicitly, so the previous references were also strong. To answer your question, when the type holding the reference (`EnvoyHTTPStreamImpl`) goes out of scope (`dealloc` is called), it releases its references to those objects